File: couchbase/logic/analytics.py
#  Copyright 2016-2022. Couchbase, Inc.
#  All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License")
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import json
from datetime import timedelta
from enum import Enum
from typing import (Any,
                    Dict,
                    List,
                    Optional,
                    Union)

from couchbase._utils import to_microseconds
from couchbase.exceptions import ErrorMapper, InvalidArgumentException
from couchbase.exceptions import exception as CouchbaseBaseException
from couchbase.logic.options import AnalyticsOptionsBase
from couchbase.options import AnalyticsOptions, UnsignedInt64
from couchbase.pycbc_core import analytics_query
from couchbase.serializer import DefaultJsonSerializer, Serializer
from couchbase.tracing import CouchbaseSpan


class AnalyticsScanConsistency(Enum):
    """
    For use with :attr:`~.AnalyticsQuery.consistency`, will allow cached
    values to be returned. This will improve performance but may not
    reflect the latest data in the server.
    """
    NOT_BOUNDED = "not_bounded"
    REQUEST_PLUS = "request_plus"


class AnalyticsStatus(Enum):
    """
    Represents the status of an analytics query.
    """
    RUNNING = "running"
    SUCCESS = "success"
    ERRORS = "errors"
    COMPLETED = "completed"
    STOPPED = "stopped"
    TIMEOUT = "timeout"
    CLOSED = "closed"
    FATAL = "fatal"
    ABORTED = "aborted"
    UNKNOWN = "unknown"


class AnalyticsProblem(object):
    def __init__(self, raw):
        self._raw = raw

    def code(self) -> int:
        return self._raw.get("code", None)

    def message(self) -> str:
        return self._raw.get("message", None)


class AnalyticsWarning(AnalyticsProblem):
    def __init__(self, analytics_warning):
        super().__init__(analytics_warning)

    def __repr__(self):
        return "AnalyticsWarning:{}".format(super()._raw)


class AnalyticsError(AnalyticsProblem):
    def __init__(self, analytics_error):
        super().__init__(analytics_error)

    def __repr__(self):
        return "AnalyticsError:{}".format(super()._raw)


class AnalyticsMetrics(object):

    def __init__(self, raw  # type: Dict[str, Any]
                 ) -> None:
        self._raw = raw

    def elapsed_time(self) -> timedelta:
        us = self._raw.get("elapsed_time") / 1000
        return timedelta(microseconds=us)

    def execution_time(self) -> timedelta:
        us = self._raw.get("execution_time") / 1000
        return timedelta(microseconds=us)

    def result_count(self) -> UnsignedInt64:
        return UnsignedInt64(self._raw.get("result_count", 0))

    def result_size(self) -> UnsignedInt64:
        return UnsignedInt64(self._raw.get("result_size", 0))

    def error_count(self) -> UnsignedInt64:
        return UnsignedInt64(self._raw.get("error_count", 0))

    def processed_objects(self) -> UnsignedInt64:
        return UnsignedInt64(self._raw.get("processed_objects", 0))

    def warning_count(self) -> UnsignedInt64:
        return UnsignedInt64(self._raw.get("warning_count", 0))

    def __repr__(self):
        return "AnalyticsMetrics:{}".format(self._raw)


class AnalyticsMetaData:
    def __init__(self, raw  # type: Dict[str, Any]
                 ) -> None:
        if raw is not None:
            self._raw = raw.get('metadata', None)
            sig = self._raw.get('signature', None)
            if sig is not None:
                self._raw['signature'] = json.loads(sig)
        else:
            self._raw = None

    def request_id(self) -> str:
        return self._raw.get("request_id", None)

    def client_context_id(self) -> str:
        return self._raw.get("client_context_id", None)

    def status(self) -> AnalyticsStatus:
        return AnalyticsStatus[self._raw.get("status", "unknown").upper()]

    def signature(self) -> Optional[Dict[str, Any]]:
        return self._raw.get("signature", None)

    def warnings(self) -> List[AnalyticsWarning]:
        return list(
            map(AnalyticsWarning, self._raw.get("warnings", []))
        )

    def errors(self) -> List[AnalyticsError]:
        return list(
            map(AnalyticsError, self._raw.get("errors", []))
        )

    def metrics(self) -> Optional[AnalyticsMetrics]:
        if "metrics" in self._raw:
            return AnalyticsMetrics(self._raw.get("metrics", {}))
        return None

    def __repr__(self):
        return "AnalyticsMetaData:{}".format(self._raw)


class AnalyticsQuery:

    _VALID_OPTS = {
        'timeout': {'timeout': lambda x: x},
        'read_only': {'readonly': lambda x: x},
        'scan_consistency': {'consistency': lambda x: x.value},
        'client_context_id': {'client_context_id': lambda x: x},
        'priority': {'priority': lambda x: x},
        'query_context': {'query_context': lambda x: x},
        'serializer': {'serializer': lambda x: x},
        'raw': {'raw': lambda x: x},
        'positional_parameters': {},
        'named_parameters': {},
        'span': {'span': lambda x: x}
    }

    def __init__(self, query, *args, **kwargs):

        self._adhoc = True
        self._params = {"statement": query}
        self._raw = None
        if args:
            self._add_pos_args(*args)
        if kwargs:
            self._set_named_args(**kwargs)

    def _set_named_args(self, **kv):
        """
        Set a named parameter in the query. The named field must
        exist in the query itself.

        :param kv: Key-Value pairs representing values within the
            query. These values should be stripped of their leading
            `$` identifier.

        """
        # named_params = {}
        # for k in kv:
        #     named_params["${0}".format(k)] = json.dumps(kv[k])
        # couchbase++ wants all args JSONified
        named_params = {f'${k}': json.dumps(v) for k, v in kv.items()}

        self._params["named_parameters"] = named_params
        return self

    def _add_pos_args(self, *args):
        """
        Set values for *positional* placeholders (``$1,$2,...``)

        :param args: Values to be used
        """
        arg_array = self._params.setdefault("positional_parameters", [])
        # couchbase++ wants all args JSONified
        json_args = [json.dumps(arg) for arg in args]
        arg_array.extend(json_args)

    def set_option(self, name, value):
        """
        Set a raw option in the query. This option is encoded
        as part of the query parameters without any client-side
        verification. Use this for settings not directly exposed
        by the Python client.

        :param name: The name of the option
        :param value: The value of the option
        """
        self._params[name] = value

    @property
    def params(self):
        return self._params

    @property
    def timeout(self) -> Optional[float]:
        value = self._params.get('timeout', None)
        if not value:
            return None
        value = value[:-1]
        return float(value)

    @timeout.setter
    def timeout(self, value  # type: Union[timedelta,float,int]
                ) -> None:
        if not value:
            self._params.pop('timeout', 0)
        else:
            total_us = to_microseconds(value)
            self.set_option('timeout', total_us)

    @property
    def metrics(self):
        return self._params.get("metrics", True)

    @metrics.setter
    def metrics(self, value):
        self.set_option("metrics", value)

    @property
    def priority(self):
        return self._params.get("priority", False)

    @priority.setter
    def priority(self, value):
        self.set_option("priority", value)

    @property
    def statement(self):
        return self._params["statement"]

    @property
    def consistency(self):
        return self._params.get(
            "scan_consistency", AnalyticsScanConsistency.NOT_BOUNDED.value
        )

    @consistency.setter
    def consistency(self, value):
        self._params["scan_consistency"] = value

    @property
    def client_context_id(self) -> Optional[str]:
        return self._params.get('client_context_id', None)

    @client_context_id.setter
    def client_context_id(self, value  # type: str
                          ) -> None:
        self.set_option('client_context_id', value)

    @property
    def readonly(self):
        value = self._params.get("readonly", False)
        return value

    @readonly.setter
    def readonly(self, value):
        self._params["readonly"] = value

    @property
    def query_context(self) -> Optional[str]:
        return self._params.get('scope_qualifier', None)

    @query_context.setter
    def query_context(self, value  # type: str
                      ) -> None:
        self.set_option('scope_qualifier', value)

    @property
    def serializer(self):
        return self._params.get("serializer", None)

    @serializer.setter
    def serializer(self, value):
        if not issubclass(value.__class__, Serializer):
            raise InvalidArgumentException('Serializer should implement Serializer interface.')
        self._params["serializer"] = value

    @property
    def raw(self) -> Optional[Dict[str, Any]]:
        return self._params.get('raw', None)

    @raw.setter
    def raw(self, value  # type: Dict[str, Any]
            ) -> None:
        if not isinstance(value, dict):
            raise TypeError("Raw option must be of type Dict[str, Any].")
        for k in value.keys():
            if not isinstance(k, str):
                raise TypeError("key for raw value must be str")
        raw_params = {f'{k}': json.dumps(v) for k, v in value.items()}
        self.set_option('raw', raw_params)

    @property
    def span(self) -> Optional[CouchbaseSpan]:
        return self._params.get('span', None)

    @span.setter
    def span(self, value  # type: CouchbaseSpan
             ):
        if not issubclass(value.__class__, CouchbaseSpan):
            raise InvalidArgumentException('Span should implement CouchbaseSpan interface.')
        self.set_option('span', value)

    @classmethod
    def create_query_object(cls, statement, *options, **kwargs):
        # lets make a copy of the options, and update with kwargs...
        opt = AnalyticsOptions()
        # TODO: is it possible that we could have [AnalyticsOptions, AnalyticsOptions, ...]??
        #       If so, why???
        opts = list(options)
        for o in opts:
            if isinstance(o, (AnalyticsOptions, AnalyticsOptionsBase)):
                opt = o
                opts.remove(o)
        args = opt.copy()
        args.update(kwargs)

        # now lets get positional parameters.  Actual positional
        # params OVERRIDE positional_parameters
        positional_parameters = args.pop("positional_parameters", [])
        if opts and len(opts) > 0:
            positional_parameters = opts

        # now the named parameters.  NOTE: all the kwargs that are
        # not VALID_OPTS must be named parameters, and the kwargs
        # OVERRIDE the list of named_parameters
        new_keys = list(filter(lambda x: x not in cls._VALID_OPTS, args.keys()))
        named_parameters = args.pop("named_parameters", {})
        for k in new_keys:
            named_parameters[k] = args[k]

        query = cls(statement, *positional_parameters, **named_parameters)
        # now lets try to setup the options.
        # but for now we will use the existing N1QLQuery.  Could be we can
        # add to it, etc...

        # default to True on analytics metrics
        query.metrics = args.get("metrics", True)

        for k, v in ((k, args[k]) for k in (args.keys() & cls._VALID_OPTS)):
            for target, transform in cls._VALID_OPTS[k].items():
                setattr(query, target, transform(v))
        return query


class AnalyticsRequestLogic:
    def __init__(self,
                 connection,
                 query_params,
                 row_factory=lambda x: x,
                 **kwargs
                 ):

        self._connection = connection
        self._query_params = query_params
        self.row_factory = row_factory
        self._streaming_result = None
        self._default_serializer = kwargs.pop('default_serializer', DefaultJsonSerializer())
        self._serializer = None
        self._started_streaming = False
        self._streaming_timeout = kwargs.pop('streaming_timeout', None)
        self._done_streaming = False
        self._metadata = None

    @property
    def params(self) -> Dict[str, Any]:
        return self._query_params

    @property
    def serializer(self) -> Serializer:
        if self._serializer:
            return self._serializer

        serializer = self.params.get('serializer', None)
        if not serializer:
            serializer = self._default_serializer

        self._serializer = serializer
        return self._serializer

    @property
    def started_streaming(self) -> bool:
        return self._started_streaming

    @property
    def done_streaming(self) -> bool:
        return self._done_streaming

    def metadata(self):
        # @TODO:  raise if query isn't complete?
        return self._metadata

    def _set_metadata(self, analytics_response):
        if isinstance(analytics_response, CouchbaseBaseException):
            raise ErrorMapper.build_exception(analytics_response)

        self._metadata = AnalyticsMetaData(analytics_response.raw_result.get('value', None))

    def _submit_query(self, **kwargs):
        if self.done_streaming:
            return

        self._started_streaming = True
        analytics_kwargs = {
            'conn': self._connection,
        }
        analytics_kwargs.update(self.params)

        streaming_timeout = self.params.get('timeout', self._streaming_timeout)
        if streaming_timeout:
            analytics_kwargs['streaming_timeout'] = streaming_timeout

        # this is for txcouchbase...
        callback = kwargs.pop('callback', None)
        if callback:
            analytics_kwargs['callback'] = callback

        errback = kwargs.pop('errback', None)
        if errback:
            analytics_kwargs['errback'] = errback

        self._streaming_result = analytics_query(**analytics_kwargs)

    def __iter__(self):
        raise NotImplementedError(
            'Cannot use synchronous iterator, are you using `async for`?'
        )

    def __aiter__(self):
        raise NotImplementedError(
            'Cannot use asynchronous iterator.'
        )

File: src/analytics.cxx
/*
 *   Copyright 2016-2022. Couchbase, Inc.
 *   All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

#include "analytics.hxx"
#include "exceptions.hxx"
#include "result.hxx"
#include "tracing.hxx"
#include <core/analytics_scan_consistency.hxx>

couchbase::core::analytics_scan_consistency
str_to_scan_consistency_type(std::string consistency)
{
    if (consistency.compare("not_bounded") == 0) {
        return couchbase::core::analytics_scan_consistency::not_bounded;
    }
    if (consistency.compare("request_plus") == 0) {
        return couchbase::core::analytics_scan_consistency::request_plus;
    }

    // TODO: better exception
    PyErr_SetString(PyExc_ValueError, "Invalid Analytics Scan Consistency type.");
    return {};
}

PyObject*
analytics_status_to_string(couchbase::core::operations::analytics_response::analytics_status status)
{
    std::string status_str;
    switch (status) {
        case couchbase::core::operations::analytics_response::analytics_status::running: {
            status_str = "running";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::success: {
            status_str = "success";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::errors: {
            status_str = "errors";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::completed: {
            status_str = "completed";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::stopped: {
            status_str = "stopped";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::timedout: {
            status_str = "timeout";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::closed: {
            status_str = "closed";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::fatal: {
            status_str = "fatal";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::aborted: {
            status_str = "aborted";
            break;
        }
        case couchbase::core::operations::analytics_response::analytics_status::unknown:
        default: {
            status_str = "unknown";
            break;
        }
    }
    // should not be able to reach here, since this is an enum class
    return PyUnicode_FromString(status_str.c_str());
}

PyObject*
get_result_metrics(couchbase::core::operations::analytics_response::analytics_metrics metrics)
{
    PyObject* pyObj_metrics = PyDict_New();
    std::chrono::duration<unsigned long long, std::nano> int_nsec = metrics.elapsed_time;
    PyObject* pyObj_tmp = PyLong_FromUnsignedLongLong(int_nsec.count());
    if (-1 == PyDict_SetItemString(pyObj_metrics, "elapsed_time", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    int_nsec = metrics.execution_time;
    pyObj_tmp = PyLong_FromUnsignedLongLong(int_nsec.count());
    if (-1 == PyDict_SetItemString(pyObj_metrics, "execution_time", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyLong_FromUnsignedLongLong(metrics.result_count);
    if (-1 == PyDict_SetItemString(pyObj_metrics, "result_count", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyLong_FromUnsignedLongLong(metrics.result_size);
    if (-1 == PyDict_SetItemString(pyObj_metrics, "result_size", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyLong_FromUnsignedLongLong(metrics.error_count);
    if (-1 == PyDict_SetItemString(pyObj_metrics, "error_count", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyLong_FromUnsignedLongLong(metrics.processed_objects);
    if (-1 == PyDict_SetItemString(pyObj_metrics, "processed_objects", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyLong_FromUnsignedLongLong(metrics.warning_count);
    if (-1 == PyDict_SetItemString(pyObj_metrics, "warning_count", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    return pyObj_metrics;
}

PyObject*
get_result_metadata(couchbase::core::operations::analytics_response::analytics_meta_data metadata, bool include_metrics)
{
    PyObject* pyObj_metadata = PyDict_New();
    PyObject* pyObj_tmp = PyUnicode_FromString(metadata.request_id.c_str());
    if (-1 == PyDict_SetItemString(pyObj_metadata, "request_id", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = PyUnicode_FromString(metadata.client_context_id.c_str());
    if (-1 == PyDict_SetItemString(pyObj_metadata, "client_context_id", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    pyObj_tmp = analytics_status_to_string(metadata.status);
    if (-1 == PyDict_SetItemString(pyObj_metadata, "status", pyObj_tmp)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_tmp);

    if (metadata.signature.has_value()) {
        pyObj_tmp = json_decode(metadata.signature.value().c_str(), metadata.signature.value().length());
        if (-1 == PyDict_SetItemString(pyObj_metadata, "signature", pyObj_tmp)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_tmp);
    }

    PyObject* pyObj_warnings = PyList_New(static_cast<Py_ssize_t>(0));
    for (auto const& warning : metadata.warnings) {
        PyObject* pyObj_warning = PyDict_New();

        pyObj_tmp = PyLong_FromLong(warning.code);
        if (-1 == PyDict_SetItemString(pyObj_warning, "code", pyObj_tmp)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_tmp);

        pyObj_tmp = PyUnicode_FromString(warning.message.c_str());
        if (-1 == PyDict_SetItemString(pyObj_warning, "message", pyObj_tmp)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_tmp);

        if (-1 == PyList_Append(pyObj_warnings, pyObj_warning)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_warning);
    }

    if (-1 == PyDict_SetItemString(pyObj_metadata, "warnings", pyObj_warnings)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_warnings);

    PyObject* pyObj_errors = PyList_New(static_cast<Py_ssize_t>(0));
    for (auto const& error : metadata.errors) {
        PyObject* pyObj_error = PyDict_New();

        pyObj_tmp = PyLong_FromLong(error.code);
        if (-1 == PyDict_SetItemString(pyObj_error, "code", pyObj_tmp)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_tmp);

        pyObj_tmp = PyUnicode_FromString(error.message.c_str());
        if (-1 == PyDict_SetItemString(pyObj_error, "message", pyObj_tmp)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_tmp);

        if (-1 == PyList_Append(pyObj_errors, pyObj_error)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObj_error);
    }

    if (-1 == PyDict_SetItemString(pyObj_metadata, "errors", pyObj_errors)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_errors);

    if (include_metrics) {
        PyObject* pyObject_metrics = get_result_metrics(metadata.metrics);

        if (-1 == PyDict_SetItemString(pyObj_metadata, "metrics", pyObject_metrics)) {
            PyErr_Print();
            PyErr_Clear();
        }
        Py_XDECREF(pyObject_metrics);
    }

    return pyObj_metadata;
}

result*
create_result_from_analytics_response(couchbase::core::operations::analytics_response resp, bool include_metrics)
{
    PyObject* pyObj_result = create_result_obj();
    result* res = reinterpret_cast<result*>(pyObj_result);
    res->ec = resp.ctx.ec;

    PyObject* pyObj_payload = PyDict_New();

    PyObject* pyObject_metadata = get_result_metadata(resp.meta, include_metrics);
    if (-1 == PyDict_SetItemString(pyObj_payload, "metadata", pyObject_metadata)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObject_metadata);

    if (-1 == PyDict_SetItemString(res->dict, RESULT_VALUE, pyObj_payload)) {
        PyErr_Print();
        PyErr_Clear();
    }
    Py_XDECREF(pyObj_payload);

    return res;
}

void
create_analytics_result(couchbase::core::operations::analytics_response resp,
                        bool include_metrics,
                        std::shared_ptr<rows_queue<PyObject*>> rows,
                        PyObject* pyObj_callback,
                        PyObject* pyObj_errback)
{
    auto set_exception = false;
    PyObject* pyObj_exc = nullptr;
    PyObject* pyObj_args = NULL;
    PyObject* pyObj_func = NULL;
    PyObject* pyObj_callback_res = nullptr;

    PyGILState_STATE state = PyGILState_Ensure();
    if (resp.ctx.ec.value()) {
        pyObj_exc = build_exception_from_context(resp.ctx, __FILE__, __LINE__, "Error doing analytics operation.");
        // lets clear any errors
        PyErr_Clear();
        rows->put(pyObj_exc);
    } else {
        for (auto const& row : resp.rows) {
            PyObject* pyObj_row = PyBytes_FromStringAndSize(row.c_str(), row.length());
            rows->put(pyObj_row);
        }

        auto res = create_result_from_analytics_response(resp, include_metrics);
        if (res == nullptr || PyErr_Occurred() != nullptr) {
            set_exception = true;
        } else {
            // None indicates done (i.e. raise StopIteration)
            Py_INCREF(Py_None);
            rows->put(Py_None);
            rows->put(reinterpret_cast<PyObject*>(res));
        }
    }

    if (set_exception) {
        pyObj_exc = pycbc_build_exception(PycbcError::UnableToBuildResult, __FILE__, __LINE__, "Analytics operation error.");
        rows->put(pyObj_exc);
    }

    // This is for txcouchbase -- let it knows we're done w/ the analytics request
    if (pyObj_callback != nullptr) {
        pyObj_func = pyObj_callback;
        pyObj_args = PyTuple_New(1);
        PyTuple_SET_ITEM(pyObj_args, 0, PyBool_FromLong(static_cast<long>(1)));
    }

    if (pyObj_func != nullptr) {
        pyObj_callback_res = PyObject_CallObject(pyObj_func, pyObj_args);
        if (pyObj_callback_res) {
            Py_DECREF(pyObj_callback_res);
        } else {
            pycbc_set_python_exception(PycbcError::InternalSDKError, __FILE__, __LINE__, "Analytics complete callback failed.");
        }
        Py_DECREF(pyObj_args);
        Py_XDECREF(pyObj_callback);
        Py_XDECREF(pyObj_errback);
    }

    PyGILState_Release(state);
}

streamed_result*
handle_analytics_query([[maybe_unused]] PyObject* self, PyObject* args, PyObject* kwargs)
{
    // need these for all operations
    PyObject* pyObj_conn = nullptr;
    char* statement = nullptr;

    char* scan_consistency = nullptr;
    char* bucket_name = nullptr;
    char* scope_name = nullptr;
    char* scope_qualifier = nullptr;
    char* client_context_id = nullptr;

    uint64_t timeout = 0;
    // booleans, but use int to read from kwargs
    int metrics = 0;
    int readonly = 0;
    int priority = 0;

    PyObject* pyObj_raw = nullptr;
    PyObject* pyObj_named_parameters = nullptr;
    PyObject* pyObj_positional_parameters = nullptr;
    PyObject* pyObj_serializer = nullptr;
    PyObject* pyObj_callback = nullptr;
    PyObject* pyObj_errback = nullptr;
    PyObject* pyObj_row_callback = nullptr;
    PyObject* pyObj_span = nullptr;

    static const char* kw_list[] = { "conn",
                                     "statement",
                                     "bucket_name",
                                     "scope_name",
                                     "scope_qualifier",
                                     "client_context_id",
                                     "scan_consistency",
                                     "timeout",
                                     "metrics",
                                     "readonly",
                                     "priority",
                                     "named_parameters",
                                     "positional_parameters",
                                     "raw",
                                     "serializer",
                                     "callback",
                                     "errback",
                                     "row_callback",
                                     "span",
                                     nullptr };

    const char* kw_format = "O!s|sssssLiiiOOOOOOOO";
    int ret = PyArg_ParseTupleAndKeywords(args,
                                          kwargs,
                                          kw_format,
                                          const_cast<char**>(kw_list),
                                          &PyCapsule_Type,
                                          &pyObj_conn,
                                          &statement,
                                          &bucket_name,
                                          &scope_name,
                                          &scope_qualifier,
                                          &client_context_id,
                                          &scan_consistency,
                                          &timeout,
                                          &metrics,
                                          &readonly,
                                          &priority,
                                          &pyObj_named_parameters,
                                          &pyObj_positional_parameters,
                                          &pyObj_raw,
                                          &pyObj_serializer,
                                          &pyObj_callback,
                                          &pyObj_errback,
                                          &pyObj_row_callback,
                                          &pyObj_span);
    if (!ret) {
        PyErr_SetString(PyExc_ValueError, "Unable to parse arguments");
        return nullptr;
    }

    connection* conn = nullptr;
    std::chrono::milliseconds timeout_ms = couchbase::core::timeout_defaults::analytics_timeout;

    conn = reinterpret_cast<connection*>(PyCapsule_GetPointer(pyObj_conn, "conn_"));
    if (nullptr == conn) {
        PyErr_SetString(PyExc_ValueError, "passed null connection");
        return nullptr;
    }
    PyErr_Clear();

    if (0 < timeout) {
        timeout_ms = std::chrono::milliseconds(std::max(0ULL, timeout / 1000ULL));
    }

    couchbase::core::operations::analytics_request req{ statement };
    // positional parameters
    std::vector<couchbase::core::json_string> positional_parameters{};
    if (pyObj_positional_parameters && PyList_Check(pyObj_positional_parameters)) {
        size_t nargs = static_cast<size_t>(PyList_Size(pyObj_positional_parameters));
        size_t ii;
        for (ii = 0; ii < nargs; ++ii) {
            PyObject* pyOb_param = PyList_GetItem(pyObj_positional_parameters, ii);
            if (!pyOb_param) {
                // TODO:  handle this better
                PyErr_SetString(PyExc_ValueError, "Unable to parse positional argument.");
                return nullptr;
            }
            // PyList_GetItem returns borrowed ref, inc while using, decr after done
            Py_INCREF(pyOb_param);
            if (PyUnicode_Check(pyOb_param)) {
                auto res = std::string(PyUnicode_AsUTF8(pyOb_param));
                positional_parameters.push_back(couchbase::core::json_string{ std::move(res) });
            }
            //@TODO: exception if this check fails??
            Py_DECREF(pyOb_param);
            pyOb_param = nullptr;
        }
    }
    if (positional_parameters.size() > 0) {
        req.positional_parameters = positional_parameters;
    }

    // named parameters
    std::map<std::string, couchbase::core::json_string> named_parameters{};
    if (pyObj_named_parameters && PyDict_Check(pyObj_named_parameters)) {
        PyObject *pyObj_key, *pyObj_value;
        Py_ssize_t pos = 0;

        // PyObj_key and pyObj_value are borrowed references
        while (PyDict_Next(pyObj_named_parameters, &pos, &pyObj_key, &pyObj_value)) {
            std::string k;
            if (PyUnicode_Check(pyObj_key)) {
                k = std::string(PyUnicode_AsUTF8(pyObj_key));
            }
            if (PyUnicode_Check(pyObj_value) && !k.empty()) {
                auto res = std::string(PyUnicode_AsUTF8(pyObj_value));
                named_parameters.emplace(k, couchbase::core::json_string{ std::move(res) });
            }
        }
    }
    if (named_parameters.size() > 0) {
        req.named_parameters = named_parameters;
    }

    req.timeout = timeout_ms;
    // req.metrics = metrics == 1;
    req.readonly = readonly == 1;
    req.priority = priority == 1;

    if (scan_consistency) {
        req.scan_consistency = str_to_scan_consistency_type<couchbase::core::analytics_scan_consistency>(scan_consistency);
    }

    if (scope_qualifier != nullptr) {
        req.scope_qualifier = std::string(scope_qualifier);
    }

    // raw options
    std::map<std::string, couchbase::core::json_string> raw_options{};
    if (pyObj_raw && PyDict_Check(pyObj_raw)) {
        PyObject *pyObj_key, *pyObj_value;
        Py_ssize_t pos = 0;

        // PyObj_key and pyObj_value are borrowed references
        while (PyDict_Next(pyObj_raw, &pos, &pyObj_key, &pyObj_value)) {
            std::string k;
            if (PyUnicode_Check(pyObj_key)) {
                k = std::string(PyUnicode_AsUTF8(pyObj_key));
            }
            if (PyUnicode_Check(pyObj_value) && !k.empty()) {
                auto res = std::string(PyUnicode_AsUTF8(pyObj_value));
                raw_options.emplace(k, couchbase::core::json_string{ std::move(res) });
            }
        }
    }
    if (raw_options.size() > 0) {
        req.raw = raw_options;
    }
    if (nullptr != pyObj_span) {
        req.parent_span = std::make_shared<pycbc::request_span>(pyObj_span);
    }

    // PyObjects that need to be around for the cxx client lambda
    // have their increment/decrement handled w/in the callback_context struct
    // struct callback_context callback_ctx = { pyObj_callback, pyObj_errback };
    Py_XINCREF(pyObj_errback);
    Py_XINCREF(pyObj_callback);

    // timeout is always set either to default, or timeout provided in options
    streamed_result* streamed_res = create_streamed_result_obj(req.timeout.value());

    // TODO:  let the couchbase++ streaming stabilize a bit more...
    // req.row_callback = [rows = streamed_res->rows](std::string&& row) {
    //     PyGILState_STATE state = PyGILState_Ensure();
    //     PyObject* pyObj_row = PyBytes_FromStringAndSize(row.c_str(), row.length());
    //     rows->put(pyObj_row);
    //     PyGILState_Release(state);
    //     return couchbase::core::utils::json::stream_control::next_row;
    // };

    {
        Py_BEGIN_ALLOW_THREADS conn->cluster_->execute(
          req,
          [rows = streamed_res->rows, include_metrics = metrics, pyObj_callback, pyObj_errback](
            couchbase::core::operations::analytics_response resp) {
              create_analytics_result(resp, include_metrics, rows, pyObj_callback, pyObj_errback);
          });
        Py_END_ALLOW_THREADS
    }
    return streamed_res;
}
